src/main/scala/viper/silver/parser/Resolver.scala

```diff
@@ -199,56 +197,43 @@ case class TypeChecker(names: NameAnalyser) {
-         */
-        val newnode: PStmt = PMethodCall(Seq(idnuse), func, args)(p.pos)
-        check(newnode)
-
-      case PVarAssign(idnuse, rhs) =>
-        names.definition(curMember)(idnuse) match {
-          case Some(PLocalVarDecl(_, typ, _)) =>
-            check(idnuse, typ)
-            check(rhs, typ)
-          case Some(PFormalArgDecl(_, typ)) =>
-            check(idnuse, typ)
-            check(rhs, typ)
-          case _ =>
-            messages ++= FastMessaging.message(stmt, "expected variable as lhs")
+      case PAssign(targets, PCall(func, args, _)) if names.definition(curMember)(func).get.isInstanceOf[PMethod] =>
+        val PMethod(_, formalArgs, formalTargets, _, _, _) = names.definition(curMember)(func).get.asInstanceOf[PMethod]
+        formalArgs.foreach(fa=>check(fa.typ))
+        if (formalArgs.length != args.length) {
+          messages ++= FastMessaging.message(stmt, "wrong number of arguments")
+        } else {
+          if (formalTargets.length != targets.length) {
+            messages ++= FastMessaging.message(stmt, "wrong number of targets")
+          } else {
+            for ((formal, actual) <- (formalArgs zip args) ++ (formalTargets zip targets)) {
+              check(actual, formal.typ)
+            }
+          }
         }
-      case PNewStmt(target, fields) =>
-        val msg = "expected variable as lhs"
-        acceptAndCheckTypedEntity[PLocalVarDecl, PFormalArgDecl](Seq(target), msg) { (v, _) => check(v, Ref) }
-        fields foreach (_.foreach(field =>
+      case PAssign(targets, _) if targets.length != 1 =>
+        messages ++= FastMessaging.message(stmt, "expected a method call")
+      case PAssign(Seq(target), PNewExp(fieldsOpt)) =>
+        check(target, Ref)
+        fieldsOpt map (_.foreach (field =>
           names.definition(curMember)(field) match {
             case Some(PField(_, typ)) =>
               check(field, typ)
             case _ =>
               messages ++= FastMessaging.message(stmt, "expected a field as argument")
           }))
-      case PMethodCall(targets, method, args) =>
-        names.definition(curMember)(method) match {
-          case Some(PMethod(_, formalArgs, formalTargets, _, _, _)) =>
-            formalArgs.foreach(fa => check(fa.typ))
-            if (formalArgs.length != args.length) {
-              messages ++= FastMessaging.message(stmt, "wrong number of arguments")
-            } else {
-              if (formalTargets.length != targets.length) {
-                messages ++= FastMessaging.message(stmt, "wrong number of targets")
-              } else {
-                for ((formal, actual) <- (formalArgs zip args) ++ (formalTargets zip targets)) {
-                  check(actual, formal.typ)
-                }
-              }
-            }
+      case PAssign(Seq(idnuse: PIdnUse), rhs) =>
+        names.definition(curMember)(idnuse) match {
+          case Some(PFormalArgDecl(_, typ)) =>
```

Comment:
I don't understand this case. If we're assigning to a local variable, then the declaration is presumably a PLocalVarDecl and therefore different from a PFormalArgDecl, and we would land in the other case and report an error? I guess not, because then all tests would fail, but I don't get why this works.